File: native/cocos/renderer/renderer/Pass.cpp
/****************************************************************************
 Copyright (c) 2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#include "Pass.h"

RENDERER_BEGIN

Pass::Pass(const std::string& programName)
: _programName(programName)
{
    _hashName = std::hash<std::string>{}(programName);
//    RENDERER_LOGD("Pass constructor: %p", this);
}

Pass::~Pass()
{
//    RENDERER_LOGD("Pass destructor: %p", this);
}

void Pass::setCullMode(CullMode cullMode)
{
    _cullMode = cullMode;
}

void Pass::setBlend(BlendOp blendEq,
                    BlendFactor blendSrc,
                    BlendFactor blendDst,
                    BlendOp blendAlphaEq,
                    BlendFactor blendSrcAlpha,
                    BlendFactor blendDstAlpha,
                    uint32_t blendColor)
{
    _blend = true;
    _blendEq = blendEq;
    _blendSrc = blendSrc;
    _blendDst = blendDst;
    _blendAlphaEq = blendAlphaEq;
    _blendSrcAlpha = blendSrcAlpha;
    _blendDstAlpha = blendDstAlpha;
    _blendColor = blendColor;
}

void Pass::setDepth(bool depthTest, bool depthWrite, DepthFunc depthFunc)
{
    _depthTest = depthTest;
    _depthWrite = depthWrite;
    _depthFunc = depthFunc;
}

void Pass::setStencilFront(StencilFunc stencilFunc,
                           uint32_t stencilRef,
                           uint8_t stencilMask,
                           StencilOp stencilFailOp,
                           StencilOp stencilZFailOp,
                           StencilOp stencilZPassOp,
                           uint8_t stencilWriteMask)
{
    _stencilTest = true;
    _stencilFuncFront = stencilFunc;
    _stencilRefFront = stencilRef;
    _stencilMaskFront = stencilMask;
    _stencilFailOpFront = stencilFailOp;
    _stencilZFailOpFront = stencilZFailOp;
    _stencilZPassOpFront = stencilZPassOp;
    _stencilWriteMaskFront = stencilWriteMask;
}

void Pass::setStencilBack(StencilFunc stencilFunc,
                          uint32_t stencilRef,
                          uint8_t stencilMask,
                          StencilOp stencilFailOp,
                          StencilOp stencilZFailOp,
                          StencilOp stencilZPassOp,
                          uint8_t stencilWriteMask)
{
    _stencilTest = true;
    _stencilFuncBack = stencilFunc;
    _stencilRefBack = stencilRef;
    _stencilMaskBack = stencilMask;
    _stencilFailOpBack = stencilFailOp;
    _stencilZFailOpBack = stencilZFailOp;
    _stencilZPassOpBack = stencilZPassOp;
    _stencilWriteMaskBack = stencilWriteMask;
}

void Pass::copy(const Pass& pass)
{
    // blending
    _blend = pass._blend;
    _blendEq = pass._blendEq;
    _blendAlphaEq = pass._blendAlphaEq;
    _blendSrc = pass._blendSrc;
    _blendDst = pass._blendDst;
    _blendSrcAlpha = pass._blendSrcAlpha;
    _blendDstAlpha = pass._blendDstAlpha;
    _blendColor = pass._blendColor;
    
    // depth
    _depthTest = pass._depthTest;
    _depthWrite = pass._depthWrite;
    _depthFunc = pass._depthFunc;
    
    // stencil
    
    _stencilTest = pass._stencilTest;
    // front
    _stencilRefFront = pass._stencilRefFront;
    _stencilFuncFront = pass._stencilFuncFront;
    _stencilFailOpFront = pass._stencilFailOpFront;
    _stencilZFailOpFront = pass._stencilZFailOpFront;
    _stencilZPassOpFront = pass._stencilZPassOpFront;
    _stencilWriteMaskFront = pass._stencilWriteMaskFront;
    _stencilMaskFront = pass._stencilMaskFront;
    // back
    _stencilRefBack = pass._stencilRefBack;
    _stencilFuncBack = pass._stencilFuncBack;
    _stencilFailOpBack = pass._stencilFailOpBack;
    _stencilZFailOpBack = pass._stencilZFailOpBack;
    _stencilZPassOpBack = pass._stencilZPassOpBack;
    _stencilWriteMaskBack = pass._stencilWriteMaskBack;
    _stencilMaskBack = pass._stencilMaskBack;
    
    // cull mode
    _cullMode = pass._cullMode;
    
    _programName = pass._programName;
    _hashName = pass._hashName;
}

RENDERER_END

File: cocos/renderer/core/pass.js
// Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

import gfx from '../../gfx';

export default class Pass {
  constructor(name) {
    this._programName = name;

    // cullmode
    this._cullMode = gfx.CULL_BACK;

    // blending
    this._blend = false;
    this._blendEq = gfx.BLEND_FUNC_ADD;
    this._blendAlphaEq = gfx.BLEND_FUNC_ADD;
    this._blendSrc = gfx.BLEND_ONE;
    this._blendDst = gfx.BLEND_ZERO;
    this._blendSrcAlpha = gfx.BLEND_ONE;
    this._blendDstAlpha = gfx.BLEND_ZERO;
    this._blendColor = 0xffffffff;

    // depth
    this._depthTest = false;
    this._depthWrite = false;
    this._depthFunc = gfx.DS_FUNC_LESS,

    // stencil
    this._stencilTest = false;
    // front
    this._stencilFuncFront = gfx.DS_FUNC_ALWAYS;
    this._stencilRefFront = 0;
    this._stencilMaskFront = 0xff;
    this._stencilFailOpFront = gfx.STENCIL_OP_KEEP;
    this._stencilZFailOpFront = gfx.STENCIL_OP_KEEP;
    this._stencilZPassOpFront = gfx.STENCIL_OP_KEEP;
    this._stencilWriteMaskFront = 0xff;
    // back
    this._stencilFuncBack = gfx.DS_FUNC_ALWAYS;
    this._stencilRefBack = 0;
    this._stencilMaskBack = 0xff;
    this._stencilFailOpBack = gfx.STENCIL_OP_KEEP;
    this._stencilZFailOpBack = gfx.STENCIL_OP_KEEP;
    this._stencilZPassOpBack = gfx.STENCIL_OP_KEEP;
    this._stencilWriteMaskBack = 0xff;
  }

  setCullMode(cullMode) {
    this._cullMode = cullMode;
  }

  setBlend(
    blendEq = gfx.BLEND_FUNC_ADD,
    blendSrc = gfx.BLEND_ONE,
    blendDst = gfx.BLEND_ZERO,
    blendAlphaEq = gfx.BLEND_FUNC_ADD,
    blendSrcAlpha = gfx.BLEND_ONE,
    blendDstAlpha = gfx.BLEND_ZERO,
    blendColor = 0xffffffff
  ) {
    this._blend = true;
    this._blendEq = blendEq;
    this._blendSrc = blendSrc;
    this._blendDst = blendDst;
    this._blendAlphaEq = blendAlphaEq;
    this._blendSrcAlpha = blendSrcAlpha;
    this._blendDstAlpha = blendDstAlpha;
    this._blendColor = blendColor;
  }

  setDepth(
    depthTest = false,
    depthWrite = false,
    depthFunc = gfx.DS_FUNC_LESS
  ) {
    this._depthTest = depthTest;
    this._depthWrite = depthWrite;
    this._depthFunc = depthFunc;
  }

  setStencilFront(
    enabled = false,
    stencilFunc = gfx.DS_FUNC_ALWAYS,
    stencilRef = 0,
    stencilMask = 0xff,
    stencilFailOp = gfx.STENCIL_OP_KEEP,
    stencilZFailOp = gfx.STENCIL_OP_KEEP,
    stencilZPassOp = gfx.STENCIL_OP_KEEP,
    stencilWriteMask = 0xff
  ) {
    this._stencilTest = enabled;
    this._stencilFuncFront = stencilFunc;
    this._stencilRefFront = stencilRef;
    this._stencilMaskFront = stencilMask;
    this._stencilFailOpFront = stencilFailOp;
    this._stencilZFailOpFront = stencilZFailOp;
    this._stencilZPassOpFront = stencilZPassOp;
    this._stencilWriteMaskFront = stencilWriteMask;
  }

  setStencilBack(
    enabled = false,
    stencilFunc = gfx.DS_FUNC_ALWAYS,
    stencilRef = 0,
    stencilMask = 0xff,
    stencilFailOp = gfx.STENCIL_OP_KEEP,
    stencilZFailOp = gfx.STENCIL_OP_KEEP,
    stencilZPassOp = gfx.STENCIL_OP_KEEP,
    stencilWriteMask = 0xff
  ) {
    this._stencilTest = enabled;
    this._stencilFuncBack = stencilFunc;
    this._stencilRefBack = stencilRef;
    this._stencilMaskBack = stencilMask;
    this._stencilFailOpBack = stencilFailOp;
    this._stencilZFailOpBack = stencilZFailOp;
    this._stencilZPassOpBack = stencilZPassOp;
    this._stencilWriteMaskBack = stencilWriteMask;
  }
}